File: Data Structures & Algorithms/binary-tree-diameter/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0
        def dfs(node):
            # 1. base case
            if not node:
                return 0

            # 2. recursive case

            # no pre case

            # recurse
            left = dfs(node.left)
            right = dfs(node.right)

            # post case
            diameter = left + right
            nonlocal res
            res = max(res, diameter)
            height = 1 + max(left, right)

            return height

        dfs(root)
        return res

        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-6.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if(root == null)
            return root;

        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while(!queue.isEmpty()) {

            TreeNode current = queue.poll();

            // swap left and right
            TreeNode left = current.left;
            TreeNode right = current.right;

            current.left = right;
            current.right = left;
            
            if(current.left != null)
                queue.add(current.left);

            if(current.right != null)
                queue.add(current.right);
        }
        return root;
    }

}
